### PR TITLE
Homework 2 SHM

### PR DESCRIPTION
**Pull request by arafalski**
Authors: @Vicx95, @kontrybutor, @piezol, @kari0036, @anraf1001

### CODE REVIEW

File: shm/inc/alcohol.hpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    size_t getAmount() const override { return amount_; }
+    size_t getPrice() const override;
+    size_t getBasePrice() const override { return basePrice_; }
+
+    //Override from Observer
+    void nextDay() override;
+
+    bool operator==(const Cargo& cargoToCheck) const override;
+    Cargo& operator+=(size_t amount) override;
+    Cargo& operator-=(size_t amount) override;
```
Comment (nauka-programowania-MA) [inline shm/inc/alcohol.hpp:25]: This is also override from Cargo

File: shm/inc/alcohol.hpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+protected:
+    // override from Cloneable<Cargo>
+    Cargo* cloneToRawPointer() override { return new Alcohol(*this); }
```
Comment (nauka-programowania-MA) [inline shm/inc/alcohol.hpp]: Dangerous. You allocate memory and need to manage it. You can clone as shared_ptr.

File: shm/inc/cloneable.hpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+template <class T>
+class Cloneable {
+public:
+    std::shared_ptr<T> cloneToShared() {
+        return static_cast<std::shared_ptr<T>>(cloneToRawPointer());
+    }
+    std::unique_ptr<T> cloneToUnique() {
+        return static_cast<std::unique_ptr<T>>(cloneToRawPointer());
+    }
```
Comment (nauka-programowania-MA) [inline shm/inc/cloneable.hpp]: when `cloneToRawPtr` will be replaced as cloneAsSharedPtr you can ommit static_cast.

File: shm/inc/delegate.hpp [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+#include <iostream>
+
+class Delegate {
+public:
+    virtual void payCrew(size_t money) = 0;
+    virtual ~Delegate() {}
+};
```
Comment (nauka-programowania-MA) [inline shm/inc/delegate.hpp]: Delegate should not be standalone class. You can create a lot of delegates and then you have conflict with names.

File: shm/inc/fruit.hpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    bool operator==(const Cargo& cargoToCheck) const override;
+    Cargo& operator+=(size_t amount) override;
+    Cargo& operator-=(size_t amount) override;
+
+    //override from Observer
+    void nextDay() override;
+
+protected:
+    // override from Cloneable<Cargo>
+    Cargo* cloneToRawPointer() override { return new Fruit(*this); }
```
Comment (nauka-programowania-MA) [inline shm/inc/fruit.hpp]: cloneToSharedPtr

File: shm/inc/item.hpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    size_t getAmount() const override { return amount_; }
+    size_t getBasePrice() const override { return basePrice_; }
+    size_t getPrice() const override;
+
+    //override from Observer
+    void nextDay() override;
+
+    bool operator==(const Cargo& cargoToCheck) const override;
+    Cargo& operator+=(size_t amount) override;
+    Cargo& operator-=(size_t amount) override;
```
Comment (nauka-programowania-MA) [inline shm/inc/item.hpp:29]: also override from Cargo

File: shm/inc/item.hpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+protected:
+    // override from Cloneable<Cargo>
+    Cargo* cloneToRawPointer() override { return new Item(*this); }
```
Comment (nauka-programowania-MA) [inline shm/inc/item.hpp]: shared_ptr

File: shm/inc/item.hpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+private:
+    Rarity rarity_;
+
+    const std::map<Rarity, size_t> priceMultipliers_{{Rarity::common, 1},
+                                                     {Rarity::rare, 2},
+                                                     {Rarity::epic, 4},
+                                                     {Rarity::legendary, 8}};
```
Comment (nauka-programowania-MA) [inline shm/inc/item.hpp]: you can use enum number as price. Common = 1, rare = 2, epic = 4 etc...

File: shm/inc/observer.hpp [hunk 9]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+class Observer {
+public:
+    virtual void nextDay() = 0;
+    virtual ~Observer(){};
+};
```
Comment (nauka-programowania-MA) [inline shm/inc/observer.hpp]: Observer should not be as standalone class. You can create a lot of observers.

File: shm/inc/player.hpp [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    void payCrew(size_t money) override;
+
+    virtual ~Player() {}
+
+    void changeMoney(size_t amount);
+    void clearEmptyCargo() { ship_->clearEmptyCargo(); };
+
+    /* override from CargoHolder through its ship */
+    void receiveCargo(Cargo* cargo, size_t amount, CargoHolder* cargoHolder) override { ship_->receiveCargo(cargo, amount, cargoHolder); }
+    void clearEmptyCargos() override { ship_->clearEmptyCargos(); }
```
Comment (nauka-programowania-MA) [inline shm/inc/player.hpp]: to cpp. This is not setter or getter

File: shm/inc/store.hpp [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    void clearEmptyCargos() override;
+
+    /* override from Observer */
+    void nextDay();
+
+    friend std::ostream& operator<<(std::ostream& out, const Store& store);
+
+private:
+    std::vector<std::shared_ptr<Cargo>> cargo_;
+    Time* time_;
```
Comment (nauka-programowania-MA) [inline shm/inc/store.hpp:61]: You don't need pointer to Time*. You have service locator, so you can always use getTime(). What's more You can provide different time to locator, and then you will use wrong or deleted ptr.

**Review by nauka-programowania-MA — CHANGES_REQUESTED** [review]
Please fix issues and the make new pull request without 301 files. This is really hard to read.

**Review by WBigaj — COMMENTED** [review]
Are these 301 files really neccessary. It's hard to dig through it to find relevant code.

File: shm/inc/alcohol.hpp [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    //override from Cargo
+    std::string getName() const override { return name_; }
+    size_t getAmount() const override { return amount_; }
+    size_t getPrice() const override;
+    size_t getBasePrice() const override { return basePrice_; }
+    std::string oneLineDescription() const override;
+    bool operator==(const Cargo& cargoToCheck) const override;
+    Cargo& operator+=(size_t amount) override;
+    Cargo& operator-=(size_t amount) override;
+    std::shared_ptr<Cargo> getShared() override { return std::make_shared<Alcohol>(Alcohol(*this)); }
```
Comment (nauka-programowania-MA) [inline shm/inc/alcohol.hpp:26]: maybe just clone?

File: shm/inc/cargo.hpp [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <memory>
+#include <string>
+
+#include "time.hpp"
+
+class Cargo : public Time::Observer {
+public:
+    Cargo(const std::string& name, size_t amount, size_t basePrice);
+    virtual ~Cargo() = default;
```
Comment (nauka-programowania-MA) [inline shm/inc/cargo.hpp:11]: = default is also definition. Move to cpp

File: shm/inc/dryfruit.hpp [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+constexpr auto kDrynessPriceMultiplier = 3;
+}  // namespace
+
+class DryFruit : public Fruit {
+public:
+    using Fruit::Fruit;
+
+    Fruit& operator--() override;
+
+    // override from Cargo.Fruit
```
Comment (nauka-programowania-MA) [inline shm/inc/dryfruit.hpp:16]: Just Fruit

File: shm/inc/player.hpp [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+class Player : public CargoHolder,
+               public Ship::Delegate {
+public:
+    Player(Ship& ship, size_t money, size_t availableSpace);
+
+    std::shared_ptr<Ship> getShip() { return ship_; }
+    size_t getMoney() const { return money_; }
+    size_t getavailableSpace() const { return availableSpace_; }
+    size_t getSpeed() const { return ship_->getSpeed(); }
+    std::shared_ptr<Cargo> getCargo(size_t index) const { return ship_->getCargo(index); }
```
Comment (nauka-programowania-MA) [inline shm/inc/player.hpp:17]: Are You validate somewhere index?

**Review by nauka-programowania-MA — APPROVED** [review]

**Comment by nauka-programowania-MA** [comment]
{
"requirement_ids": [2568, 2568, 2568, 2571, 2571, 2571, 2569, 2570],
"points": [5, 5, 5, 8, 8, 8, 20, 6],
"users": ["Vicx95", "kontrybutor", "piezol", "kari0036", "anraf1001"]
}

**Pull request closed without merging**